test/cim/constructor_validation.py

```diff
@@ -45,34 +43,98 @@
 peu_args = (*equip_args, PowerElectronicsConnection("test_pec"), -100, 200)
 
-
 sw_kwargs = {**ce_kwargs}
 
 sw_args = ce_args
 
-
 ps_kwargs = {**sw_kwargs}
 
 ps_args = sw_args
 
+tsi_args = (
+    *io_args,
+    floats(min_value=FLOAT_MIN, max_value=FLOAT_MAX),
+    floats(min_value=FLOAT_MIN, max_value=FLOAT_MAX),
+    floats(min_value=FLOAT_MIN, max_value=FLOAT_MAX),
+    floats(min_value=FLOAT_MIN, max_value=FLOAT_MAX),
+    builds(TransformerEndInfo, mrid="test_transformer_end_info")
+)
+
+tsi_kwargs = {
+    **io_kwargs,
+    "r": floats(min_value=FLOAT_MIN, max_value=FLOAT_MAX),
+    "r0": floats(min_value=FLOAT_MIN, max_value=FLOAT_MAX),
+    "x": floats(min_value=FLOAT_MIN, max_value=FLOAT_MAX),
+    "x0": floats(min_value=FLOAT_MIN, max_value=FLOAT_MAX),
+    "transformer_end_info": builds(TransformerEndInfo, mrid="test_transformer_end_info")
+}
+
+ai_kwargs = {**io_kwargs}
+ai_args = io_args
+
+tei_args = (
+    *ai_args,
+    WindingConnection.UNKNOWN_WINDING,
+    integers(min_value=MIN_32_BIT_INTEGER, max_value=MAX_32_BIT_INTEGER),
+    integers(min_value=MIN_32_BIT_INTEGER, max_value=MAX_32_BIT_INTEGER),
+    integers(min_value=MIN_32_BIT_INTEGER, max_value=MAX_32_BIT_INTEGER),
+    integers(min_value=MIN_32_BIT_INTEGER, max_value=MAX_32_BIT_INTEGER),
+    floats(min_value=FLOAT_MIN, max_value=FLOAT_MAX),
+    integers(min_value=MIN_32_BIT_INTEGER, max_value=MAX_32_BIT_INTEGER),
+    integers(min_value=MIN_32_BIT_INTEGER, max_value=MAX_32_BIT_INTEGER),
+    integers(min_value=MIN_32_BIT_INTEGER, max_value=MAX_32_BIT_INTEGER),
+    integers(min_value=MIN_32_BIT_INTEGER, max_value=MAX_32_BIT_INTEGER),
+)
+
+tei_kwargs = {
+    "connection_kind": WindingConnection.UNKNOWN_WINDING,
+    "emergency_s": integers(min_value=MIN_32_BIT_INTEGER, max_value=MAX_32_BIT_INTEGER),
+    "end_number": integers(min_value=MIN_32_BIT_INTEGER, max_value=MAX_32_BIT_INTEGER),
+    "insulation_u": integers(min_value=MIN_32_BIT_INTEGER, max_value=MAX_32_BIT_INTEGER),
+    "phase_angle_clock": integers(min_value=MIN_32_BIT_INTEGER, max_value=MAX_32_BIT_INTEGER),
+    "r": floats(min_value=FLOAT_MIN, max_value=FLOAT_MAX),
+    "rated_s": integers(min_value=MIN_32_BIT_INTEGER, max_value=MAX_32_BIT_INTEGER),
+    "rated_u": integers(min_value=MIN_32_BIT_INTEGER, max_value=MAX_32_BIT_INTEGER),
+    "short_term_s": integers(min_value=MIN_32_BIT_INTEGER, max_value=MAX_32_BIT_INTEGER),
+    "transformer_star_impedance": integers(min_value=MIN_32_BIT_INTEGER, max_value=MAX_32_BIT_INTEGER),
+    **ai_kwargs,
+}
+
+pti_args = (
+    *ai_args,
+    lists(builds(TransformerTankInfo, **assetinfo()), max_size=2)
+)
+
+pti_kwargs = {
+    **ai_kwargs,
+    "transformer_tank_infos": lists(builds(TransformerTankInfo, **assetinfo()), max_size=2)
+}
+
+tti_args = (
+    *ai_args,
+    lists(builds(TransformerEndInfo, **assetinfo()), max_size=2)
+)
+
+tti_kwargs = {
+    **ai_kwargs,
+    "transformer_end_infos": lists(builds(TransformerEndInfo, **assetinfo()), max_size=2)
```

Comment:
```suggestion
    "transformer_end_infos": lists(builds(TransformerEndInfo, mrid="test_transformer_end_info"), max_size=2)
```

the number of combinations tends to increase dramatically when you start building types with all their properties, which then increases the test runtime massively, so just passing an mRID like you've done above is generally fine; we don't care about testing the properties of the sub-type (TransformerEndInfo), as that will get tested through the TransformerEndInfo's tests.